app/cloud.go

```diff
@@ -62,57 +121,146 @@ func (a *App) NotifySystemAdminsToUpgrade(c *request.Context, currentUserTeamID
 		return appErr
 	}
 
+	now := model.GetMillis()
+
+	data, err := a.Srv().Store.NotifyAdmin().Get(trial)
+	if err != nil {
+		return model.NewAppError("SendNotifyAdminPosts", "app.notify_admin.send_notification_post.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	data = filterNotificationData(data, func(nad *model.NotifyAdminData) bool { return string(nad.RequiredPlan) != currentSKU })
+
+	if len(data) == 0 {
+		mlog.Warn("No notification data available")
+		return nil
+	}
+
+	userBasedData := a.groupNotifyAdminByUser(data)
+	featureBasedData := a.groupNotifyAdminByFeature(data)
+	props := make(model.StringInterface)
+
 	for _, admin := range sysadmins {
 		T := i18n.GetUserTranslations(admin.Locale)
+		message := T("app.cloud.upgrade_plan_bot_message", map[string]interface{}{"UsersNum": len(userBasedData), "WorkspaceName": workspaceName})
+		if len(userBasedData) == 1 {
+			message = T("app.cloud.upgrade_plan_bot_message_single", map[string]interface{}{"UsersNum": len(userBasedData), "WorkspaceName": workspaceName}) // todo (allan): investigate if translations library can do this
+		}
+		if trial {
+			message = T("app.cloud.trial_plan_bot_message", map[string]interface{}{"UsersNum": len(userBasedData), "WorkspaceName": workspaceName})
+			if len(userBasedData) == 1 {
+				message = T("app.cloud.trial_plan_bot_message_single", map[string]interface{}{"UsersNum": len(userBasedData), "WorkspaceName": workspaceName})
+			}
+		}
+
 		channel, appErr := a.GetOrCreateDirectChannel(c, systemBot.UserId, admin.Id)
 		if appErr != nil {
 			mlog.Warn("Error getting direct channel", mlog.Err(appErr))
 			continue
 		}
 
 		post := &model.Post{
-			Message:   T("api.cloud.upgrade_plan_bot_message", map[string]any{"TeamName": team.Name}),
+			Message:   message,
 			UserId:    systemBot.UserId,
 			ChannelId: channel.Id,
 			Type:      fmt.Sprintf("%sup_notification", model.PostCustomTypePrefix), // webapp will have to create renderer for this custom post type
+
 		}
 
+		props["requested_features"] = featureBasedData
+		props["trial"] = trial
+		post.SetProps(props)
+
 		_, appErr = a.CreatePost(c, post, channel, false, true)
 		if appErr != nil {
 			mlog.Warn("Error creating post", mlog.Err(appErr))
 			continue
 		}
 	}
 
-	// mark as done for current user until end of cool off period
-	out, err := json.Marshal(&model.AdminNotificationUserInfo{
-		LastUserIDToNotify:        userId,
-		LastNotificationTimestamp: model.GetMillis(),
-	})
+	a.FinishSendAdminNotifyPost(trial, now)
+
+	return nil
+}
+
+func (a *App) UserAlreadyNotifiedOnRequiredFeature(user string, feature model.MattermostPaidFeature) bool {
+	data, err := a.Srv().Store.NotifyAdmin().GetDataByUserIdAndFeature(user, feature)
 	if err != nil {
-		mlog.Warn("Unable to Marshal", mlog.Err(err))
+		return true // any error should flag as already notified to avoid data corruption like duplicates
```

Comment:
My view is we should avoid duplicates. The first notification (if any), already serves the purpose the extra notification would.

Reply:
Yes, but how can we be certain there was a first notification? We don't have the value to know because of the error so it may have not have been notified before and defaulting here to `true` then the request won't be sent and we would lose it

Reply:
Alright. I have updated the code to default to `false`